### PR TITLE
refactor(e2e): change `*Node` to string in keys of some maps

### PR DESCRIPTION
**Pull request by sergio-mena**
Partially addresses #2458

Change the `*Node` as keys of some maps in `e2e` and replace them by the node's name.

Having pointers as keys of a map brings about subtleties (how comparison is done) that we don't need. Code is much clearer with node names, even if it takes a bit longer to execute some loops (we need to look up the nodes). But this is test code, so clarity and robustness comes before performance of the testing code.

I decided to tackle this refactor in an different PR, rather than piggybacking it to #2283

---

#### PR checklist

- [x] Tests written/updated
- ~[ ] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)~
- ~[ ] Updated relevant documentation (`docs/` or `spec/`) and code comments~

### CODE REVIEW

File: test/e2e/tests/validator_test.go [hunk 1]
```diff
@@ -178,10 +178,11 @@ func (s *validatorSchedule) Increment(heights int64) {
...
 }
 
-func makeVals(valMap map[*e2e.Node]int64) []*types.Validator {
+func makeVals(testnet *e2e.Testnet, valMap map[string]int64) []*types.Validator {
 	vals := make([]*types.Validator, 0, len(valMap))
-	for node, power := range valMap {
-		vals = append(vals, types.NewValidator(node.PrivvalKey.PubKey(), power))
+	for valName, power := range valMap {
+		validator := testnet.LookupNode(valName)
+		vals = append(vals, types.NewValidator(validator.PrivvalKey.PubKey(), power))
```
Comment (andynog) [inline test/e2e/tests/validator_test.go:185]: the `LookupNode` can return `nil`, should test for that and only append in that case ? not sure if there are corner cases that can make it return `nil` but just highlighting the fact here, but it would panic if `validator` is `nil` too

**Review by andynog — APPROVED** [review]
lgtm, had one comment but I think it is irrelevant to the current logic so feel free to disregard.

**Review by melekes — APPROVED** [review]

File: test/e2e/tests/validator_test.go (hunk 1 above)
Comment (alesforz) [reply test/e2e/tests/validator_test.go:185]: This applies to other parts of the refactorings above as well.

File: test/e2e/pkg/testnet.go [hunk 2]
```diff
@@ -81,8 +81,8 @@ type Testnet struct {
 	Dir  string
 
 	IP               *net.IPNet
-	Validators       map[*Node]int64
-	ValidatorUpdates map[int64]map[*Node]int64
+	Validators       map[string]int64
```
Comment (cason) [inline test/e2e/pkg/testnet.go:84]: Since the type of `Validators` here is now identical to in `pkg/manifest.go` we don't even need to keep it here.

Namely, you could only rename `ValidatorsMap` to `Validators` in `Manifest` and remove this field from here.

File: test/e2e/pkg/testnet.go [hunk 3]
```diff
@@ -81,8 +81,8 @@ type Testnet struct {
...
+	ValidatorUpdates map[int64]map[string]int64
```
Comment (cason) [inline test/e2e/pkg/testnet.go:85]: Same comment as above.

File: test/e2e/pkg/testnet.go (hunk 2 above)
Comment (cason) [reply test/e2e/pkg/testnet.go:84]: See https://github.com/cometbft/cometbft/pull/3964 for more details.

**Review by cason — COMMENTED** [review]
Following https://github.com/cometbft/cometbft/pull/3964, we should remove the two refactored fields from `Testnet` type, since they match the version on `Manifest`.

File: test/e2e/tests/validator_test.go (hunk 1 above)
Comment (sergio-mena) [reply test/e2e/tests/validator_test.go:185]: Good point folks. I'll revisit the PR with this in mind

File: test/e2e/pkg/testnet.go (hunk 3 above)
Comment (sergio-mena) [reply test/e2e/pkg/testnet.go:85]: `manifest.ValidatorUpdatesMap` is of type `map[string]map[string]int64`, whereas `testnet.ValidatorUpdates` is of type`map[int64]map[string]int64`.
I'm sure there's a way to turn `manifest.ValidatorUpdatesMap` into the same type as `testnet.ValidatorUpdates`, and thus get rid of the latter. But as this is not trivial, I'd rather leave it out of this PR.

File: test/e2e/tests/validator_test.go (hunk 1 above)
Comment (sergio-mena) [reply test/e2e/tests/validator_test.go:195]: Addressed

File: test/e2e/pkg/testnet.go (hunk 2 above)
Comment (sergio-mena) [reply test/e2e/pkg/testnet.go]: Addressed

Comment (cason) [reply test/e2e/pkg/testnet.go]: We can rename it back to `Validators`, see https://github.com/cometbft/cometbft/pull/4166

File: test/e2e/pkg/testnet.go [hunk 4]
```diff
@@ -74,15 +74,14 @@ const (
 // Testnet represents a single testnet.
 // It includes all fields from the associated Manifest instance.
 type Testnet struct {
-	Manifest
+	*Manifest
```
Comment (cason) [inline test/e2e/pkg/testnet.go:77]: Are you sure this doesn't break anything?

File: test/e2e/runner/setup.go [hunk 5]
```diff
@@ -150,9 +150,13 @@ func MakeGenesis(testnet *e2e.Testnet) (types.GenesisDoc, error) {
 	if testnet.PbtsUpdateHeight == -1 {
 		genesis.ConsensusParams.Feature.PbtsEnableHeight = testnet.PbtsEnableHeight
 	}
-	for validator, power := range testnet.Validators {
+	for valName, power := range *testnet.Manifest.ValidatorsMap {
+		validator := testnet.LookupNode(valName)
+		if validator == nil {
+			return types.GenesisDoc{}, fmt.Errorf("unknown validator %q for genesis doc", valName)
```
Comment (cason) [inline test/e2e/runner/setup.go:156]: Should never happen, right?

File: test/e2e/runner/setup.go [hunk 6]
```diff
@@ -425,8 +432,12 @@ func MakeAppConfig(node *e2e.Node) ([]byte, error) {
 		validatorUpdates := map[string]map[string]int64{}
 		for height, validators := range node.Testnet.ValidatorUpdates {
 			updateVals := map[string]int64{}
-			for node, power := range validators {
-				updateVals[base64.StdEncoding.EncodeToString(node.PrivvalKey.PubKey().Bytes())] = power
+			for valName, power := range validators {
+				validator := node.Testnet.LookupNode(valName)
+				if validator == nil {
+					return nil, fmt.Errorf("unknown validator %q for validator updates in testnet, height %d", valName, height)
```
Comment (cason) [inline test/e2e/runner/setup.go:438]: Should never happen too, should we panic?

**Review by cason — APPROVED** [review]
Legit.

**Comment by cason** [comment]
Consider merging https://github.com/cometbft/cometbft/pull/4166 into this.

File: test/e2e/pkg/testnet.go (hunk 4 above)
Comment (sergio-mena) [reply test/e2e/pkg/testnet.go:77]: It was breaking when not used as a pointer, because we would copy some fields (non-pointer fields) but not others. A total mess.
As we are now modifying the Manifest down below in this function, those modifications didn't propagate (hence I needed to turn it into a pointer)

File: test/e2e/runner/setup.go (hunk 5 above)
Comment (sergio-mena) [reply test/e2e/runner/setup.go:156]: Yes, but as suggested by @alesforz and @andynog , it's cleaner code if we check for `nil`.

File: test/e2e/runner/setup.go (hunk 6 above)
Comment (sergio-mena) [reply test/e2e/runner/setup.go:438]: I'd rather return the error and have the callsite(s) decide whether they want to panic or not
